### PR TITLE
Implement Web Storage

### PR DESCRIPTION
**Pull request by sh7dm**
- [x] `Storage` interface
- [x] `sessionStorage`
- [ ] `localStorage`
#1657.

### CODE REVIEW

**Comment by kitsonk** [comment]
These need to align to what is in [`lib.dom.d.ts`](https://github.com/Microsoft/TypeScript/blob/f944ed6d72577240b276e8bb03cc402033e1da6b/lib/lib.dom.d.ts#L14568).  There is a single global interface `Storage` and constructor function (`window.Storage`) which needs to be newable.  Also `sessionStorage instanceof Storage === true`.

The same goes for `localStorage`.  You are going to have to manage whether you persist the data internally without change the API.  I would recommend exporting `Storage`, `sessionStorage`, and `localStorage` from `storage` and not `new` them up in `globals.ts` but rather `new` them up inside the module.  For the `localStorage` you would when persist the changes.

With the values (and possibly the keys) I think we need some sort of runtime checking (or coercion), because people will expect it if they are using just JavaScript.  For example, this works without throwing an error:

```
sessionStorage.setItem("foo", 1);
```

It will return always return a string of course.  So I would say that you would want to simply do a `String(value)` when setting a value.

@ry at the moment we don't have a concept of `window.onunload` within Deno, and we are going to need something like that so that `localStorage` can be serialised and persisted to the local file system.

**Comment by aduh95** [comment]
`sessionStorage` is supposed to be readonly [as per the specs](https://html.spec.whatwg.org/multipage/webstorage.html#dom-sessionstorage). Should we try to have that in Deno as well?

**Comment by sh7dm** [comment]
@ry maybe transfer this to deno_std

File: js/storage.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  private _storage: Record<string, string> = {};
+
+  get length(): number {
+    return Object.keys(this._storage).length;
+  }
+
+  key(index: number): string {
+    return Object.keys(this._storage)[index];
+  }
+  getItem(key: string): string {
```
Comment (zekth) [inline js/storage.ts]: Would it be revelant to return an empty string instead of an undefined value in this case?

Comment (sh7dm) [reply js/storage.ts]: @zekth I don't think so. Maybe user put `""` there? `undefined` will let user see that it's... undefined!

Comment (zekth) [reply js/storage.ts]: Both solution are good, it's more like a philosophy. Some APIs returns empty strings, others undefined as the return type is a string. Was just wondering.

Comment (aduh95) [inline js/storage.ts]:
```suggestion
  getItem(key: string): string | null {
```

File: js/session_storage_test.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  assertEqual(sessionStorage.length, 0);
+});
+
+test(function key() {
+  sessionStorage.setItem("test", "John Doe");
+  sessionStorage.setItem("test1", "Jane Doe");
+  assertEqual(sessionStorage.key(0), "test");
+  assertEqual(sessionStorage.key(1), "test1");
+  sessionStorage.clear();
+  assertEqual(sessionStorage.key(0), undefined);
```
Comment (aduh95) [inline js/session_storage_test.ts]:
```suggestion
  assertEqual(sessionStorage.key(0), null);
```

File: js/session_storage_test.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  assertEqual(sessionStorage.getItem("test"), undefined);
+  sessionStorage.clear(); // Clear for next tests
+});
+
+test(function clear() {
+  sessionStorage.setItem("test", "John Doe");
+  assertEqual(sessionStorage.getItem("test"), "John Doe");
+  assertEqual(sessionStorage.length, 1);
+  sessionStorage.clear();
+  assertEqual(sessionStorage.getItem("test"), undefined);
```
Comment (aduh95) [inline js/session_storage_test.ts]:
```suggestion
  assertEqual(sessionStorage.getItem("test"), null);
```

File: js/session_storage_test.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  assertEqual(sessionStorage.getItem("test"), "John Doe");
+  sessionStorage.clear(); // Clear for next tests
+});
+
+test(function removeItems() {
+  assertEqual(sessionStorage.getItem("test"), undefined);
+  sessionStorage.setItem("test", "John Doe");
+  assertEqual(sessionStorage.getItem("test"), "John Doe");
+  sessionStorage.removeItem("test");
+  assertEqual(sessionStorage.getItem("test"), undefined);
```
Comment (aduh95) [inline js/session_storage_test.ts]:
```suggestion
  assertEqual(sessionStorage.getItem("test"), null);
```

File: js/session_storage_test.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+test(function storeItems() {
+  assertEqual(sessionStorage.getItem("test"), undefined);
+  sessionStorage.setItem("test", "John Doe");
+  assertEqual(sessionStorage.getItem("test"), "John Doe");
+  sessionStorage.clear(); // Clear for next tests
+});
+
+test(function removeItems() {
+  assertEqual(sessionStorage.getItem("test"), undefined);
```
Comment (aduh95) [inline js/session_storage_test.ts]:
```suggestion
  assertEqual(sessionStorage.getItem("test"), null);
```

File: js/session_storage_test.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import { test, assertEqual } from "./test_util.ts";
+
+test(function storeItems() {
+  assertEqual(sessionStorage.getItem("test"), undefined);
```
Comment (aduh95) [inline js/session_storage_test.ts]:
```suggestion
  assertEqual(sessionStorage.getItem("test"), null);
```

File: js/storage.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  setItem(key: string, value: string): void;
+  removeItem(key: string): void;
+  clear(): void;
+}
+
+export class SessionStorage implements Storage {
+  private _storage: Record<string, string> = {};
+
+  get length(): number {
+    return Object.keys(this._storage).length;
```
Comment (aduh95) [inline js/storage.ts]: IMHO using a `Map` instead of an `Object` would make this way more efficient.

File: js/globals.ts [hunk 8]
```diff
@@ -96,4 +97,8 @@ export type TextDecoder = textEncoding.TextDecoder;
 
 window.performance = new performanceUtil.Performance();
 
+export type Storage = storage.Storage;
+export type SessionStorage = storage.SessionStorage;
+window.sessionStorage = new storage.SessionStorage();
```
Comment (aduh95) [inline js/globals.ts]: This should be a readonly property.

File: js/globals.ts [hunk 9]
```diff
@@ -96,4 +97,8 @@ export type TextDecoder = textEncoding.TextDecoder;
 
 window.performance = new performanceUtil.Performance();
 
+export type Storage = storage.Storage;
+export type SessionStorage = storage.SessionStorage;
```
Comment (aduh95) [inline js/globals.ts]: Do we need to export this type? It is not part of the HTML spec.

**Comment by bartlomieju** [comment]
@sh7dm will you update this PR?

**Comment by ry** [comment]
I'm still interested in adding built-in web storage support, but this needs more work and has now grown stale. Thanks @sh7dm for getting the ball rolling, I think we'll refer back to this in the future.

**Pull request closed without merging**